src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -106,41 +390,313 @@ public function request(string $method, string $url, array $options = []): Respo
+     * @param array<string, string|string[]> $responseHeaders the HTTP response headers
+     * @param array<string, string|true>     $cacheControl    an array of parsed Cache-Control directives
+     *
+     * @return bool true if the response is cacheable, false otherwise
+     */
+    private function isServerResponseCacheable(int $statusCode, array $requestHeaders, array $responseHeaders, array $cacheControl): bool
     {
-        if ($this->client instanceof ResetInterface) {
-            $this->client->reset();
+        // no-store => skip caching
+        if (isset($cacheControl['no-store'])) {
+            return false;
         }
+
+        if (
+            $this->sharedCache
+            && !isset($cacheControl['public']) && !isset($cacheControl['s-maxage']) && !isset($cacheControl['must-revalidate'])
+            && isset($requestHeaders['authorization'])
+        ) {
+            return false;
+        }
+
+        if ($this->sharedCache && isset($cacheControl['private'])) {
+            return false;
+        }
+
+        // Conditionals require an explicit expiration
+        if (\in_array($statusCode, self::CONDITIONALLY_CACHEABLE_STATUS_CODES, true)) {
+            return $this->hasExplicitExpiration($responseHeaders, $cacheControl);
+        }
+
+        return \in_array($statusCode, self::CACHEABLE_STATUS_CODES, true);
+    }
+
+    /**
+     * Checks if the response has an explicit expiration.
+     *
+     * This function will return true if the response has an explicit expiration
+     * time specified in the headers or in the Cache-Control directives,
+     * false otherwise.
+     *
+     * @param array<string, string|string[]> $headers      the HTTP response headers
+     * @param array<string, string|true>     $cacheControl an array of parsed Cache-Control directives
+     *
+     * @return bool true if the response has an explicit expiration, false otherwise
+     */
+    private function hasExplicitExpiration(array $headers, array $cacheControl): bool
+    {
+        return isset($headers['expires'])
+            || ($this->sharedCache && isset($cacheControl['s-maxage']))
+            || isset($cacheControl['max-age']);
+    }
+
+    /**
+     * Creates a MockResponse object from cached data.
+     *
+     * This function constructs a MockResponse from the cached data, including
+     * the original request method, URL, and options, as well as the cached
+     * response headers and content. The constructed MockResponse is then
+     * returned.
+     *
+     * @param string $key        the cache key for the response
+     * @param array  $cachedData the cached data for the response
+     * @param string $method     the original request method
+     * @param string $url        the original request URL
+     * @param array  $options    the original request options
+     *
+     * @return MockResponse the constructed MockResponse object
```

Comment:
the descriptions for the return value and the parameters is useless IMO as names are already clear enough (and then, the tags themselves become unnecessary for most of them thanks to the native signature)

Reply:
adressed in https://github.com/symfony/symfony/commit/b3154e1c1ecf2ad2b1c7b33e95549af3c79e3f4b